apps/admin/lib/utils/fetcher.ts

```diff
@@ -0,0 +1,123 @@
+type ApiResponse = Response & { data?: any };
+
+type RequestInterceptor = (
+  options: RequestInit,
+) => RequestInit | Promise<RequestInit>;
+type ResponseInterceptor = (
+  response: ApiResponse,
+) => ApiResponse | Promise<ApiResponse>;
+
+class Fetcher {
+  private baseUrl: string;
+  private defaultHeaders: HeadersInit;
+  private requestInterceptors: RequestInterceptor[];
+  private responseInterceptors: ResponseInterceptor[];
+
+  constructor({ baseUrl = "", defaultHeaders = {} } = {}) {
+    this.baseUrl = baseUrl;
+    this.defaultHeaders = defaultHeaders;
+    this.requestInterceptors = [];
+    this.responseInterceptors = [];
+  }
+
+  setBaseUrl(baseUrl: string) {
+    this.baseUrl = baseUrl;
+  }
+
+  setDefaultHeaders(headers: HeadersInit) {
+    this.defaultHeaders = headers;
+  }
+
+  addRequestInterceptor(interceptor: RequestInterceptor) {
+    this.requestInterceptors.push(interceptor);
+  }
+
+  addResponseInterceptor(interceptor: ResponseInterceptor) {
+    this.responseInterceptors.push(interceptor);
+  }
+
+  private async interceptRequest(options: RequestInit): Promise<RequestInit> {
+    options.headers = { ...this.defaultHeaders, ...options.headers };
+
+    for (const interceptor of this.requestInterceptors) {
+      options = (await interceptor(options)) || options;
+    }
+
+    return options;
+  }
+
+  private async interceptResponse(response: Response): Promise<Response> {
+    for (const interceptor of this.responseInterceptors) {
+      response = (await interceptor(response)) || response;
+    }
+
+    return response;
+  }
+
+  private async parseJsonResponse(response: Response): Promise<any> {
+    const contentType = response.headers.get("Content-Type") || "";
+
+    if (contentType.includes("application/json")) {
+      return response.json();
+    }
+
+    return response.text();
+  }
+
+  async request(url: string, options: RequestInit = {}): Promise<any> {
+    options = await this.interceptRequest(options);
+
+    const fullUrl = this.baseUrl + url;
+
+    let response: ApiResponse = await fetch(fullUrl, options);
+    response = await this.interceptResponse(response);
+    response.data = await this.parseJsonResponse(response);
+
+    return response;
+  }
+
+  get(url: string, options: RequestInit = {}): Promise<ApiResponse> {
+    return this.request(url, { ...options, method: "GET" });
+  }
```

Comment:
p4;
axios 에서 제공하는 params 기능도 구현해볼 수 있다면 추가해보는 게 어떨까유? request 쪽에서 공통으로 처리해주면 좋을 것 같네욤
```suggestion
  get(url: string, options: RequestInit = {}, params?: Record<string, any>): Promise<ApiResponse> {
   const queryString = new URLSearchParams(params).toString();
    return this.request(`${url}${queryString}`,  { ...options, method: "GET" });
  }
```

Reply:
넵 좋아요! 반영해뒀습니다